newton/_src/solvers/implicit_mpm/solver_implicit_mpm.py

```diff
@@ -15,245 +15,378 @@
+    return phi(s) * inv_cell_volume
 
-    ground_height: float
-    """Y-coordinate of the ground"""
 
-    ground_normal: wp.vec3
-    """Normal of the ground"""
+@fem.integrand
+def integrate_collider_fraction(
+    s: fem.Sample,
+    domain: fem.Domain,
+    phi: fem.Field,
+    sdf: fem.Field,
+    inv_cell_volume: float,
+):
+    return phi(s) * wp.where(sdf(s) <= 0.0, inv_cell_volume, 0.0)
 
 
-@wp.func
-def collision_sdf(x: wp.vec3, collider: Collider):
-    ground_sdf = wp.dot(x, collider.ground_normal) - collider.ground_height
+@fem.integrand
+def integrate_mass(
+    s: fem.Sample,
+    phi: fem.Field,
+    domain: fem.Domain,
+    inv_cell_volume: float,
+    particle_density: wp.array(dtype=float),
+    particle_flags: wp.array(dtype=wp.int32),
+):
+    density = wp.where(
+        particle_flags[s.qp_index] & newton.ParticleFlags.ACTIVE, particle_density[s.qp_index], _INFINITY
+    )
+    return phi(s) * density * inv_cell_volume
 
-    min_sdf = ground_sdf
-    sdf_grad = collider.ground_normal
-    sdf_vel = wp.vec3(0.0)
-    collider_id = int(_GROUND_COLLIDER_ID)
 
-    # Find closest collider
-    for m in range(collider.meshes.shape[0]):
-        mesh = collider.meshes[m]
+@fem.integrand
+def integrate_velocity(
+    s: fem.Sample,
+    domain: fem.Domain,
+    u: fem.Field,
+    velocities: wp.array(dtype=wp.vec3),
+    dt: float,
+    gravity: wp.vec3,
+    inv_cell_volume: float,
+    particle_density: wp.array(dtype=float),
+    particle_flags: wp.array(dtype=wp.int32),
+):
+    vel_adv = velocities[s.qp_index]
 
-        query = wp.mesh_query_point_sign_normal(mesh, x, collider.query_max_dist)
+    vel_adv = wp.where(
+        particle_flags[s.qp_index] & newton.ParticleFlags.ACTIVE,
+        particle_density[s.qp_index] * (vel_adv + dt * gravity),
+        _INFINITY * vel_adv,
+    )
+    return wp.dot(u(s), vel_adv) * inv_cell_volume
 
-        if query.result:
-            cp = wp.mesh_eval_position(mesh, query.face, query.u, query.v)
 
-            offset = x - cp
-            d = wp.length(offset) * query.sign
-            sdf = d - collider.thicknesses[m]
+@fem.integrand
+def integrate_velocity_apic(
+    s: fem.Sample,
+    domain: fem.Domain,
+    u: fem.Field,
+    velocity_gradients: wp.array(dtype=wp.mat33),
+    inv_cell_volume: float,
+    particle_density: wp.array(dtype=float),
+    particle_flags: wp.array(dtype=wp.int32),
+):
+    # APIC velocity prediction
+    node_offset = domain(fem.at_node(u, s)) - domain(s)
+    vel_apic = velocity_gradients[s.qp_index] * node_offset
 
-            if sdf < min_sdf:
-                min_sdf = sdf
-                if wp.abs(d) < 0.0001:
-                    sdf_grad = wp.mesh_eval_face_normal(mesh, query.face)
-                else:
-                    sdf_grad = wp.normalize(offset) * query.sign
+    vel_adv = (
+        wp.where(particle_flags[s.qp_index] == newton.ParticleFlags.ACTIVE, particle_density[s.qp_index], _INFINITY)
```

Comment:
It's better to use bitwise comparison here to future-proof the particle flags in case we add more combinations:
```py
bool(particle_flags[particle_index] & ParticleFlags.ACTIVE)
```